### PR TITLE
Okhttp3 initial implementation

### PR DESCRIPTION
**Pull request by domesticmouse**
Here's the initial sketch of converting to OkHttp3. 

No rush on this review, I'm looking to land this early to mid Q3.

### CODE REVIEW

**Comment by domesticmouse** [comment]
This will fix https://github.com/googlemaps/google-maps-services-java/issues/148

**Comment by domesticmouse** [comment]
PTAL @markmcd @samthor

File: src/main/java/com/google/maps/DirectionsApi.java [hunk 1]
```diff
@@ -30,21 +30,20 @@
  * text strings (e.g. "Chicago, IL" or "Darwin, NT, Australia") or as latitude/longitude
  * coordinates. The Directions API can return multi-part directions using a series of waypoints.
  *
- * <p>See <a href="https://developers.google.com/maps/documentation/directions/intro">documentation</a>.
```
Comment (samthor) [inline src/main/java/com/google/maps/DirectionsApi.java]: It's not a big deal but please note that this file has nothing to do with the change.

File: src/main/java/com/google/maps/GeoApiContext.java [hunk 2]
```diff
@@ -62,49 +83,40 @@
...
-  }
+    <T, R extends ApiResponse<T>> PendingResult<T> handlePost(String hostName, String url,
+        String payload,
+        String userAgent, Class<R> clazz,
+        FieldNamingPolicy fieldNamingPolicy, long errorTimeout,
+        Integer maxRetries,
+        ExceptionsAllowedToRetry exceptionsAllowedToRetry);
+
+    interface Builder {
+
```
Comment (samthor) [inline src/main/java/com/google/maps/GeoApiContext.java]: why spacing vs not-spacing previously?

File: src/main/java/com/google/maps/GeoApiContext.java [hunk 3]
```diff
@@ -62,49 +83,40 @@
...
-    this.requestHandler = requestHandler;
-    this.exceptionsAllowedToRetry.add(OverQueryLimitException.class);
   }
 
+
   <T, R extends ApiResponse<T>> PendingResult<T> get(ApiConfig config, Class<? extends R> clazz,
-                                                     Map<String, String> params) {
+      Map<String, String> params) {
     if (channel != null && !channel.isEmpty() && !params.containsKey("channel")) {
       params.put("channel", channel);
```
Comment (samthor) [inline src/main/java/com/google/maps/GeoApiContext.java]: Is it weird to modify the params map of the caller? In the other `get()` below, you just modify the actual query.

File: src/main/java/com/google/maps/GeoApiContext.java [hunk 4]
```diff
@@ -150,7 +162,8 @@ public GeoApiContext(RequestHandler requestHandler) {
       }
     }
 
-    // Channel can be supplied per-request or per-context. We prioritize it from the request, so if it's not provided there, provide it here
+    // Channel can be supplied per-request or per-context. We prioritize it from the request,
+    // so if it's not provided there, provide it here
     if (!channelSet && channel != null && !channel.isEmpty()) {
       query.append("&channel=").append(channel);
```
Comment (samthor) [inline src/main/java/com/google/maps/GeoApiContext.java]: what if the query is empty?

File: src/main/java/com/google/maps/GeoApiContext.java [hunk 5]
```diff
@@ -236,145 +251,191 @@ private void checkContext(boolean canUseClientId) {
...
-    if (allowedToRetry) {
-      exceptionsAllowedToRetry.add(exception);
-    } else {
-      exceptionsAllowedToRetry.remove(exception);
+    /**
+     * Sets the default connect timeout for new connections. A value of 0 means no timeout.
+     *
+     * @see java.net.URLConnection#setConnectTimeout(int)
+     */
+
```
Comment (samthor) [inline src/main/java/com/google/maps/GeoApiContext.java]: extra space for some reason?

File: src/main/java/com/google/maps/GeoApiContext.java [hunk 6]
```diff
@@ -236,145 +251,191 @@ private void checkContext(boolean canUseClientId) {
...
+     *
+     * @see java.net.URLConnection#setReadTimeout(int)
+     */
+    public Builder readTimeout(long timeout, TimeUnit unit) {
+      builder.readTimeout(timeout, unit);
+      return this;
+    }
+
+    /**
+     * Sets the default write timeout for new connections. A value of 0 means no timeout.
```
Comment (samthor) [inline src/main/java/com/google/maps/GeoApiContext.java]: no `@see` here?

File: src/main/java/com/google/maps/GeoApiContext.java [hunk 7]
```diff
@@ -236,145 +251,191 @@ private void checkContext(boolean canUseClientId) {
...
+     */
+    public Builder queryRateLimit(int maxQps) {
+      builder.queriesPerSecond(maxQps);
+      return this;
+    }
+
+    /**
+     * Allows specific API exceptions to be retried or not retried.
+     */
+    public Builder toggleifExceptionIsAllowedToRetry(Class<? extends ApiException> exception,
```
Comment (samthor) [inline src/main/java/com/google/maps/GeoApiContext.java]: toggleIfException ... (fix camelcase)

Comment (samthor) [inline src/main/java/com/google/maps/GeoApiContext.java]: I might also call this "set". Toggle implies.. on/off automagically.

File: src/main/java/com/google/maps/OkHttpRequestHandler.java [hunk 8]
```diff
@@ -77,43 +74,59 @@ public OkHttpRequestHandler() {
...
-    client.setWriteTimeout(timeout, unit);
-  }
-
-  @Override
-  public void setQueriesPerSecond(int maxQps) {
-    dispatcher.setMaxRequests(maxQps);
-    dispatcher.setMaxRequestsPerHost(maxQps);
-    rateLimitExecutorService.setQueriesPerSecond(maxQps);
-  }
+    LOG.info("Request: {}", hostName + url);
```
Comment (samthor) [inline src/main/java/com/google/maps/OkHttpRequestHandler.java]: remove before release?

File: src/main/java/com/google/maps/OkHttpRequestHandler.java [hunk 9]
```diff
@@ -77,43 +74,59 @@ public OkHttpRequestHandler() {
...
+    LOG.info("Body: {}", body);
 
-  @Override @Deprecated
-  public void setQueriesPerSecond(int maxQps, int minimumInterval) {
-    // Instead of using a minimumInterval between requests, we are using a warm up period
-    // on the RateLimiter in the rateLimitExecutorService to prevent flooding the back end
-    // with requests.
-    LOG.warn("OkHttpRequestHandler#setQueriesPerSecond(int,int) deprecated, ignoring minimumInterval");
-    this.setQueriesPerSecond(maxQps);
+    return new OkHttpPendingResult<T, R>(req, client, clazz, fieldNamingPolicy, errorTimeout, maxRetries, exceptionsAllowedToRetry);
```
Comment (samthor) [inline src/main/java/com/google/maps/OkHttpRequestHandler.java]: line length?

**Comment by domesticmouse** [comment]
@samthor I'm happy to split this PR if it would help.

**Comment by samthor** [comment]
If it's just specific commits that's fine. Please link me to more, if there's more?

File: src/main/java/com/google/maps/DirectionsApi.java (hunk 1 above)
Comment (domesticmouse) [inline src/main/java/com/google/maps/DirectionsApi.java]: ack

**Comment by domesticmouse** [comment]
Hi @googlebot the patches not authored by me were pulled from master on the main project, to bring this patch up to date.

File: src/main/java/com/google/maps/OkHttpRequestHandler.java (hunk 8 above)
Comment (domesticmouse) [inline src/main/java/com/google/maps/OkHttpRequestHandler.java]: Done

File: src/main/java/com/google/maps/GeoApiContext.java (hunk 2 above)
Comment (domesticmouse) [inline src/main/java/com/google/maps/GeoApiContext.java]: I hit the whole code base with [google-java-format](https://github.com/google/google-java-format) to bring it into line with our code format standards. I can do that as a separate PR on the master branch and regnerate this branch to bring the noise down.

File: src/main/java/com/google/maps/GeoApiContext.java (hunk 3 above)
Comment (domesticmouse) [inline src/main/java/com/google/maps/GeoApiContext.java]: This isn't my code, but I did sign off on bringing it into the code base. At this point I think I'll stick a TODO on it and think about how best to clean it up.

File: src/main/java/com/google/maps/GeoApiContext.java (hunk 6 above)
Comment (domesticmouse) [inline src/main/java/com/google/maps/GeoApiContext.java]: I'm curious what I should be referring to with a `@see` here.

**Comment by samthor** [comment]
Just that the above method had a @see but this one does not :)

File: src/main/java/com/google/maps/GeoApiContext.java (hunk 7 above)
Comment (domesticmouse) [inline src/main/java/com/google/maps/GeoApiContext.java]: Agreed, done

File: src/main/java/com/google/maps/OkHttpRequestHandler.java (hunk 9 above)
Comment (domesticmouse) [inline src/main/java/com/google/maps/OkHttpRequestHandler.java]: I suspect the reformat caught this one.

**Comment by domesticmouse** [comment]
I'm unsure of the history on these `@see` lines are, as the linkage between calling this method and the underlying timeout is indirect. It may make more sense to delete them from `readTimeout` and `connectTimeout`, as neither of these methods are implemented for the GAE Java7 implementation.

The Occam's razor reason for the lack of `@see` on setWriteTimeout is that `java.net.URLConnection` doesn't contain a setWriteTimeout method...

**Pull request merged**